lib/spack/spack/build_systems/python.py

```diff
@@ -233,13 +236,42 @@ def update_external_dependencies(self):
             if "python" in self.spec.root:
                 python = self.spec.root["python"]
             else:
-                python = spack.spec.Spec("python")
-                repo = spack.repo.path.repo_for_pkg(python)
-                python.namespace = repo.namespace
-                python._mark_concrete()
-                python.external_path = self.prefix
+                python = self.get_external_python_for_prefix()
+                if not python.concrete:
+                    repo = spack.repo.path.repo_for_pkg(python)
+                    python.namespace = repo.namespace
+                    python._mark_concrete()
+                    python.external_path = self.prefix
             self.spec.add_dependency_edge(python, ("build", "link", "run"))
 
+    def get_external_python_for_prefix(self):
+        python_externals_installed = [
+            s for s in spack.store.db.query("python") if s.prefix == self.prefix
+        ]
+        if python_externals_installed:
+            return python_externals_installed[0]
+
+        python_external_config = spack.config.get("packages:python:externals", [])
+        python_externals_configured = [
+            spack.spec.Spec(item["spec"])
+            for item in python_external_config
+            if item["prefix"] == self.prefix
+        ]
+        if python_externals_configured:
+            return python_externals_configured[0]
```

Comment:
In my env I have an external python which looks like:

```
externals:
- spec: python@3.7.2
```

The result is that the spec `python@3.7.2` is added to the DAG, and it does not have details added during concretization, possibly because it is a child of an external.

After installing that and doing another `concretize -f`, this spec is picked up by the db check (on line 249 here) and later there is a failure because of insufficient spec details.

```
File ".../spack/lib/spack/spack/solver/asp.py", line 625, in solve
  setup.setup(self, specs, reuse=reuse)
File ".../spack/lib/spack/spack/solver/asp.py", line 1989, in setup
  self._facts_from_concrete_spec(reusable_spec, possible)
File ".../spack/lib/spack/spack/solver/asp.py", line 1910, in _facts_from_concrete_spec
  self.possible_oses.add(dep.os)
File ".../spack/lib/spack/spack/spec.py", line 4699, in os
  return self.architecture.os
```

i.e. the spec is missing `arch=...`.

A couple notes:

* we could assign spec details ourselves but this would result in multiple instances of the Python spec in the same dag
  * I'll furthermore note that right now `traverse.py` maintains `visited` using `id` as the key by default, so even if the specs are `==` according to `Spec.__eq__`, `Package.extendee_spec` can fail its internal assertion
* Earlier I suggested concretizing the python spec but (a) it runs into the same potential issue as the prior point and (b) you mentioned that running a concretization can be slow because of the setup time

I'm becoming more convinced that this type of coordination needs to occur at the group-of-specs level, i.e. that this loop in `asp.py`:

```
        for root in self._specs.values():
             for dep in root.traverse():
                 if dep.external:
                     dep.package.update_external_dependencies()
```

should not do the patching individually: it should search all of them for a python, and then attach it to any that need it, so they all have the same one.

Reply:
We can definitely find the platform/os for the python spec easily, but the target is actually quite hard -- we don't have any way to know how conservatively it was compiled

Doing this at the multiple-package level may be fine, but doesn't solve the problem in the case of a single spec which doesn't have any other dependencies on python. We need a more general solution than that.

Reply:
Got it there are two problems then:

* Consistency: solved by patching all the specs together rather than individually
* Choosing the right target: if we have to make a Python spec up, why not just use whatever `spack arch` would return? Or we could fall back to that if the user didn't set target on the external Python (or the external Python extension)